### ISSUE TITLE
The masking results for yoloe-26x-seg show significant differences between Torch and TensorRT.

### ISSUE THREAD

**Issue by Fafa-DL** [ultralytics/ultralytics#23326]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

Predict

### Bug

i use the export code below to get `.engin` file
```
model = YOLO("yoloe-26x-seg.pt")  # or select yoloe-26s/m-seg.pt for different sizes
names = ["keyboard"]
model.set_classes(names, model.get_text_pe(names))
model.export(format="engine")
```

**the result of torch is :**

<img width="1280" height="720" alt="Image" src="https://github.com/user-attachments/assets/7d1b139d-09ab-4364-93bf-3e738f7b8779" />

**the result of trt is :**

<img width="1280" height="720" alt="Image" src="https://github.com/user-attachments/assets/ab410429-f2c9-4a08-9838-70015118b1b8" />

It's worth noting that the trt mask result moves with the camera, while the bbox result remains normal.

whats wrong? thx

### Environment

Ultralytics 8.4.6 🚀 Python-3.10.19 torch-2.5.1+cu121 CUDA:0 (NVIDIA GeForce RTX 4090 D, 24068MiB)
Setup complete ✅ (32 CPUs, 31.1 GB RAM, 340.5/915.3 GB disk)

OS                     Linux-6.8.0-90-generic-x86_64-with-glibc2.35
Environment            Linux
Python                 3.10.19
Install                pip
Path                   /home/a/miniconda3/envs/py310/lib/python3.10/site-packages/ultralytics
RAM                    31.06 GB
Disk                   340.5/915.3 GB
CPU                    Intel Core i9-14900K
CPU count              32
GPU                    NVIDIA GeForce RTX 4090 D, 24068MiB
GPU count              1
CUDA                   12.1

numpy                  ✅ 1.26.0>=1.23.0
matplotlib             ✅ 3.10.7>=3.3.0
opencv-python          ✅ 4.11.0.86>=4.6.0
pillow                 ✅ 11.0.0>=7.1.2
pyyaml                 ✅ 6.0.3>=5.3.1
requests               ✅ 2.32.5>=2.23.0
scipy                  ✅ 1.15.3>=1.4.1
torch                  ✅ 2.5.1+cu121>=1.8.0
torch                  ✅ 2.5.1+cu121!=2.4.0,>=1.8.0; sys_platform == "win32"
torchvision            ✅ 0.20.1+cu121>=0.9.0
psutil                 ✅ 5.9.8>=5.8.0
polars                 ✅ 1.36.1>=0.20.0
ultralytics-thop       ✅ 2.0.18>=2.0.18
{'OS': 'Linux-6.8.0-90-generic-x86_64-with-glibc2.35', 'Environment': 'Linux', 'Python': '3.10.19', 'Install': 'pip', 'Path': '/home/a/miniconda3/envs/py310/lib/python3.10/site-packages/ultralytics', 'RAM': '31.06 GB', 'Disk': '340.5/915.3 GB', 'CPU': 'Intel Core i9-14900K', 'CPU count': 32, 'GPU': 'NVIDIA GeForce RTX 4090 D, 24068MiB', 'GPU count': 1, 'CUDA': '12.1', 'Package Info': {'numpy': '✅ 1.26.0>=1.23.0', 'matplotlib': '✅ 3.10.7>=3.3.0', 'opencv-python': '✅ 4.11.0.86>=4.6.0', 'pillow': '✅ 11.0.0>=7.1.2', 'pyyaml': '✅ 6.0.3>=5.3.1', 'requests': '✅ 2.32.5>=2.23.0', 'scipy': '✅ 1.15.3>=1.4.1', 'torch': '✅ 2.5.1+cu121!=2.4.0,>=1.8.0; sys_platform == "win32"', 'torchvision': '✅ 0.20.1+cu121>=0.9.0', 'psutil': '✅ 5.9.8>=5.8.0', 'polars': '✅ 1.36.1>=0.20.0', 'ultralytics-thop': '✅ 2.0.18>=2.0.18'}}

### Minimal Reproducible Example

```
# trt
model = YOLO("yoloe-26x-seg.engine")
results = model.track(frame, persist=True)
for result in results:
        if result.masks is not None:
            masks = result.masks.data.cpu().numpy()  # (N, H, W)
            boxes = result.boxes.xyxy.cpu().numpy()  # (N, 4)

# torch
model = YOLO("yoloe-26x-seg.pt") 
names = ["robots"]
model.set_classes(names, model.get_text_pe(names))
```

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Fafa-DL, thank you for your detailed report and for your interest in Ultralytics 🚀! This is an automated response to help us triage faster—an Ultralytics engineer will also assist soon.

We recommend a visit to the [Docs](https://docs.ultralytics.com/) where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it. You included a snippet (thank you!)—to make it a true MRE, please also include:
-Exact export command + full console output for `model.export(format="engine")`
-A tiny input (single image/video or a few frames) and the exact pre/post-processing used to create `frame`
-Exact inference call used for Torch vs TensorRT (same `imgsz`, `conf`, `iou`, `device`, etc.) and the full console logs
-Whether you are using FP16/INT8 in export, and any TensorRT version details

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo26"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by Fafa-DL** [comment]
the export log

```
WARNING ⚠️ TensorRT requires GPU export, automatically assigning device=0
Ultralytics 8.4.6 🚀 Python-3.10.19 torch-2.5.1+cu121 CUDA:0 (NVIDIA GeForce RTX 4090 D, 24068MiB)

PyTorch: starting from 'yoloe-26x-seg.pt' with input shape (1, 3, 640, 640) BCHW and output shape(s) ((1, 300, 38), (1, 32, 160, 160)) (163.1 MB)

ONNX: starting export with onnx 1.16.1 opset 17...
/home/a/miniconda3/envs/py310/lib/python3.10/site-packages/torch/onnx/symbolic_opset9.py:5385: UserWarning: Exporting aten::index operator of advanced indexing in opset 17 is achieved by combination of multiple ONNX operators, including Reshape, Transpose, Concat, and Gather. If indices include negative values, the exported graph will produce incorrect results.
  warnings.warn(
ONNX: slimming with onnxslim 0.1.82...
ONNX: export success ✅ 2.1s, saved as 'yoloe-26x-seg.onnx' (239.7 MB)

TensorRT: starting export with TensorRT 10.8.0.43...
[01/20/2026-12:18:24] [TRT] [I] [MemUsageChange] Init CUDA: CPU -1, GPU +0, now: CPU 1351, GPU 1959 (MiB)
[01/20/2026-12:18:26] [TRT] [I] [MemUsageChange] Init builder kernel library: CPU +2772, GPU +447, now: CPU 4324, GPU 2406 (MiB)
[01/20/2026-12:18:26] [TRT] [I] ----------------------------------------------------------------
[01/20/2026-12:18:26] [TRT] [I] Input filename:   yoloe-26x-seg.onnx
[01/20/2026-12:18:26] [TRT] [I] ONNX IR version:  0.0.8
[01/20/2026-12:18:26] [TRT] [I] Opset version:    17
[01/20/2026-12:18:26] [TRT] [I] Producer name:    pytorch
[01/20/2026-12:18:26] [TRT] [I] Producer version: 2.5.1
[01/20/2026-12:18:26] [TRT] [I] Domain:           
[01/20/2026-12:18:26] [TRT] [I] Model version:    0
[01/20/2026-12:18:26] [TRT] [I] Doc string:       
[01/20/2026-12:18:26] [TRT] [I] ----------------------------------------------------------------
TensorRT: input "images" with shape(1, 3, 640, 640) DataType.FLOAT
TensorRT: output "output0" with shape(1, 300, 38) DataType.FLOAT
TensorRT: output "output1" with shape(1, 32, 160, 160) DataType.FLOAT
TensorRT: building FP32 engine as yoloe-26x-seg.engine
[01/20/2026-12:18:26] [TRT] [I] Local timing cache in use. Profiling results in this builder pass will not be stored.
[01/20/2026-12:19:02] [TRT] [I] Compiler backend is used during engine build.
[01/20/2026-12:19:42] [TRT] [I] [GraphReduction] The approximate region cut reduction algorithm is called.
[01/20/2026-12:19:42] [TRT] [I] Detected 1 inputs and 2 output network tensors.
[01/20/2026-12:19:43] [TRT] [I] Total Host Persistent Memory: 984240 bytes
[01/20/2026-12:19:43] [TRT] [I] Total Device Persistent Memory: 109056 bytes
[01/20/2026-12:19:43] [TRT] [I] Max Scratch Memory: 19662848 bytes
[01/20/2026-12:19:43] [TRT] [I] [BlockAssignment] Started assigning block shifts. This will take 499 steps to complete.
[01/20/2026-12:19:43] [TRT] [I] [BlockAssignment] Algorithm ShiftNTopDown took 30.2586ms to assign 17 blocks to 499 nodes requiring 118733824 bytes.
[01/20/2026-12:19:43] [TRT] [I] Total Activation Memory: 118732800 bytes
[01/20/2026-12:19:43] [TRT] [I] Total Weights Memory: 314526016 bytes
[01/20/2026-12:19:43] [TRT] [I] Compiler backend is used during engine execution.
[01/20/2026-12:19:43] [TRT] [I] Engine generation completed in 76.538 seconds.
[01/20/2026-12:19:43] [TRT] [I] [MemUsageStats] Peak memory usage of TRT CPU/GPU memory allocators: CPU 20 MiB, GPU 305 MiB
TensorRT: export success ✅ 81.1s, saved as 'yoloe-26x-seg.engine' (308.4 MB)

Export complete (81.6s)
Results saved to /home/a/zane/workspace/py_scripts
Predict:         yolo predict task=segment model=yoloe-26x-seg.engine imgsz=640 
Validate:        yolo val task=segment model=yoloe-26x-seg.engine imgsz=640 data={'train': {'grounding_data': [{'img_path': '../datasets/flickr/full_images/', 'json_file': '../datasets/flickr/annotations/final_flickr_separateGT_train_segm.json'}, {'img_path': '../datasets/mixed_grounding/gqa/images', 'json_file': '../datasets/mixed_grounding/annotations/final_mixed_train_no_coco_segm.json'}, {'img_path': '../datasets/Objects365v1/images/train', 'json_file': '../datasets/Objects365v1/annotations/objects365_train_segm.json'}]}, 'val': {'yolo_data': ['/home/louis/ultra_louis_work/datasets/lvis.yaml']}}  
Visualize:       https://netron.app

```

**Comment by Y-T-G** [comment]
@ShuaiLYU Can you check this?

**Comment by ShuaiLYU** [comment]
The question has been closed by the user.

**Comment by Fafa-DL** [comment]
> The question has been closed by the user.

@ShuaiLYU i found the problem, the shape of results[0].masks.data of .engine and .pt is diff

```
masks = results[0].masks.data
print(f"Frame {frame_count}: Masks shape: {masks.shape}")

# .pt
<<< Frame 49: Masks shape: torch.Size([1, 384, 640]) 

# .engine
<<< Frame 49: Masks shape: torch.Size([1, 640, 640]) 
```

How should this situation be handled? thx

**Comment by ShuaiLYU** [comment]
> > The question has been closed by the user.
> 
> [@ShuaiLYU](https://github.com/ShuaiLYU) i found the problem, the shape of results[0].masks.data of .engine and .pt is diff
> 
> ```
> masks = results[0].masks.data
> print(f"Frame {frame_count}: Masks shape: {masks.shape}")
> 
> # .pt
> <<< Frame 49: Masks shape: torch.Size([1, 384, 640]) 
> 
> # .engine
> <<< Frame 49: Masks shape: torch.Size([1, 640, 640]) 
> ```
> 
> How should this situation be handled? thx

may I know what is the size of input image ?

**Comment by Fafa-DL** [comment]
@ShuaiLYU 1280*720

**Comment by ShuaiLYU** [comment]
Just a quick check: YOLO uses letterboxing to fit input images to 640×640 (model input size) and maps predictions back to the original image size in post-processing. Can you confirm the TensorRT model has the exact same post-processing logic here?

**Comment by Fafa-DL** [comment]
@ShuaiLYU I discovered that `results[0].plot()` can automatically map the results from `.enging/.pt` back to their original format. I checked its implementation, and the core of it is `ultralytics.utils.scale_masks`. I now refer to this implementation and have solved the problem.
